apps/aecore/lib/aecore/chain/chain_state.ex

```diff
@@ -23,34 +24,48 @@ defmodule Aecore.Chain.ChainState do
   def calculate_and_validate_chain_state!(txs, chainstate, block_height) do
     txs
     |> Enum.reduce(chainstate, fn tx, chainstate ->
-      apply_transaction_on_state!(tx, chainstate, block_height)
+      case apply_transaction_on_state(tx, chainstate, block_height) do
+        {:ok, updated_chainstate} ->
+          updated_chainstate
+
+        {:error, _reason} ->
+          chainstate
+      end
     end)
   end
 
-  @spec apply_transaction_on_state!(SignedTx.t(), chainstate(), integer()) :: chainstate()
-  def apply_transaction_on_state!(%SignedTx{data: data} = tx, chainstate, block_height) do
-    cond do
-      SignedTx.is_coinbase?(tx) ->
-        receiver_state = Map.get(chainstate.accounts, data.payload.receiver, Account.empty())
-        new_receiver_state = SignedTx.reward(data, block_height, receiver_state)
+  @spec apply_transaction_on_state(SignedTx.t(), chainstate(), integer()) :: chainstate()
+  def apply_transaction_on_state(
+        %{data: %{sender: nil, payload: %{receiver: receiver}} = data, signature: nil} = tx,
+        %{accounts: accounts} = chainstate,
+        block_height
+      ) do
+    receiver_state = Map.get(accounts, receiver, Account.empty())
+    new_receiver_state = SignedTx.reward(data, block_height, receiver_state)
 
-        new_accounts_state =
-          Map.put(chainstate.accounts, data.payload.receiver, new_receiver_state)
+    new_accounts_state = Map.put(accounts, receiver, new_receiver_state)
 
-        Map.put(chainstate, :accounts, new_accounts_state)
-
-      data.sender != nil ->
-        if SignedTx.is_valid?(tx) do
-          DataTx.process_chainstate!(data, chainstate)
-        else
-          throw({:error, "Invalid transaction"})
-        end
+    {:ok, Map.put(chainstate, :accounts, new_accounts_state)}
+  end
 
-      true ->
-        throw({:error, "Invalid transaction"})
+  def apply_transaction_on_state(%{data: %{sender: sender}} = tx, chainstate, block_height)
+      when not is_nil(sender) do
```

Comment:
Would not it be better if the guard was `when is_binary(sender)` ?
Because `{}` is not nil to and in the same way the third match of function is unreachable.